File: jmlparser-jml-tools/src/main/java/com/github/jmlparser/smt/solver/SolverAnswer.java
package com.github.jmlparser.smt.solver;

import com.github.jmlparser.smt.model.SAtom;
import com.github.jmlparser.smt.model.SExpr;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.ArrayList;
import java.util.List;

/**
 * @author Alexander Weigl
 * @version 1 (08.08.22)
 */
public class SolverAnswer {
    private final List<SExpr> answers;
    private int currentPos = 0;

    public SolverAnswer(List<SExpr> answers) {
        this.answers = answers;
    }

    public List<SExpr> getAnswers() {
        return answers;
    }

    public SolverAnswer expectSat() {
        return expectSymbol("sat");
    }

    public SolverAnswer expectUnsat() {
        return expectSymbol("unsat");
    }

    public SolverAnswer expectUnknown() {
        return expectSymbol("unknown");
    }

    public SolverAnswer expectSymbol(String symbol) {
        if (!isSymbol(symbol)) {
            throw new RuntimeException("Unexpected symbol");
        }
        return this;
    }

    public boolean isSymbol(String symbol) {
        return symbol.equals(((SAtom) peek()).getValue());
    }

    public SExpr peek() {
        return answers.get(currentPos);
    }

    public void consume() {
        currentPos++;
    }

    public List<String> consumeErrors() {
        List<String> seq = new ArrayList<>();
        while (currentPos < answers.size()) {
            if (isError()) {
                seq.add(getErrorMessage());
                consume();
            } else {
                break;
            }
        }
        return seq;
    }

    private String getErrorMessage() {
        return peek().asList().get(1).asSymbolValue();
    }

    private boolean isError() {
        try {
            return peek().asList().get(0).asSymbolValue().equals("error");
        } catch (ClassCastException e) {
            return false;
        }
    }

    @Override
    public String toString() {
        StringWriter sw = new StringWriter();
        PrintWriter pw = new PrintWriter(sw);
        answers.forEach(a -> {
            a.appendTo(pw);
            pw.println();
        });
        return sw.toString();
    }
}

File: tools/smt/src/main/kotlin/io/github/jmltoolkit/smt/solver/SolverAnswer.kt
package io.github.jmltoolkit.smt.solver

import io.github.jmltoolkit.smt.model.SAtom
import io.github.jmltoolkit.smt.model.SExpr
import java.io.PrintWriter
import java.io.StringWriter
import java.util.function.Consumer

/**
 * @author Alexander Weigl
 * @version 1 (08.08.22)
 */
class SolverAnswer(private val answers: List<SExpr>) {
    private var currentPos = 0
    fun expectSat(): SolverAnswer {
        return expectSymbol("sat")
    }

    fun expectUnsat(): SolverAnswer {
        return expectSymbol("unsat")
    }

    fun expectUnknown(): SolverAnswer {
        return expectSymbol("unknown")
    }

    fun expectSymbol(symbol: String): SolverAnswer {
        if (!isSymbol(symbol)) {
            throw RuntimeException("Unexpected symbol")
        }
        return this
    }

    fun isSymbol(symbol: String): Boolean {
        return symbol == (peek() as SAtom).value
    }

    fun peek(): SExpr {
        return answers[currentPos]
    }

    fun consume() {
        currentPos++
    }

    fun consumeErrors(): List<String> {
        val seq: MutableList<String> = ArrayList()
        while (currentPos < answers.size) {
            if (isError) {
                seq.add(errorMessage)
                consume()
            } else {
                break
            }
        }
        return seq
    }

    private val errorMessage: String
        get() = peek().asList().get(1).asSymbolValue()

    private val isError: Boolean
        get() = try {
            peek().asList().get(0).asSymbolValue().equals("error")
        } catch (e: ClassCastException) {
            false
        }

    override fun toString(): String {
        val sw = StringWriter()
        val pw = PrintWriter(sw)
        answers.forEach(Consumer { a: SExpr ->
            a.appendTo(pw)
            pw.println()
        })
        return sw.toString()
    }
}
